pkg/services/ngalert/state/cache.go

```diff
@@ -307,16 +286,38 @@ func (c *cache) recordMetrics(metrics *metrics.State) {
 	}
 }
 
-// if duplicate labels exist, keep the value from the first set
+// mergeLabels returns the combined labels from both arguments, keeping
+// labels from the first argument where there are duplicates. It does not
+// change any of its arguments, instead returning a copy of them.
 func mergeLabels(a, b data.Labels) data.Labels {
-	newLbs := make(data.Labels, len(a)+len(b))
+	result := make(data.Labels, len(a)+len(b))
 	for k, v := range a {
-		newLbs[k] = v
+		result[k] = v
 	}
 	for k, v := range b {
-		if _, ok := newLbs[k]; !ok {
-			newLbs[k] = v
+		if _, ok := result[k]; !ok {
+			result[k] = v
+		}
+	}
+	return result
+}
+
+// mergeLabelsKeepDuplicates returns the combined labels from both
+// arguments, keeping labels from the first argument where there are
+// duplicates; and all duplicates from the second argument. It does
+// not change any of its arguments, instead returning a copy of them.
+func mergeLabelsKeepDuplicates(a, b data.Labels) (result data.Labels, duplicates data.Labels) {
+	result = make(data.Labels, len(a)+len(b))
+	duplicates = make(data.Labels)
+	for k, v := range a {
+		result[k] = v
+	}
+	for k, v := range b {
+		if _, ok := result[k]; !ok {
+			result[k] = v
+		} else {
+			duplicates[k] = v
```

Comment:
I think we should add to duplicates only those labels that have the same key but different value

Reply:
This is why I do not approve the PR right away. I think we should not populate dupes map with key\values that are exact dupes, only when there are two different values for the same key